---
title: "N. Crassa SRP Pipeline Report"
author: "Tristan Sones-Dykes"
date: "22/09/23"
output: html_document
---

```{r setup, include=FALSE}
library(here)
# grab functions from src
source(here("src", "signal_peptides.r"))
source(here("src", "hydrophobicity.r"))
source(here("src", "motif_analysis.r"))

# define paths
protein_path <- here("data", "Proteins", "NC_first_60.fasta")
RNA_path <- here("data", "RNA", "AN_999.fasta")
out_name <- "N_Crassa_report"

# read in AA strings
AA_stringset <- readAAStringSet(protein_path)

# grab protein names
protein_names <- names(AA_stringset)

# run signalP and Phobius on protein sequences
signalp_output <- signalp(protein_path)
phobius_output <- r_phobius(protein_path)

# run motif analysis on 5'UTRs
motif_results <- count_motifs(RNA_path, "CNYTCNYT")

# combine dfs
combined_df <- signalp_output %>%
    full_join(phobius_output, by = "seqid") %>%
    left_join(motif_results, by = "seqid")
```

### Summary
In this doc we use SignalP and Phobius to find signal peptides and transmembrane regions in protein sequences,
then use the Kyte-Doolittle hydrophobicity scale to find a compound hydrophobicity score
to separate secreted proteins into ones that use the signal recognition particle (SRP) and 
those that don't assuming they form a bimodal distribution. Then use motif analysis to find out if Ssd-1 significantly targets one 
of these groups.

## SignalP and Hydrophobicity 
After running SignalP and Phobius, we can use the Kyte-Doolittle hydrophobicity scale to
find the compound hydrophobicity score for each protein with a signal peptide, then find
the underlying bimodal distribution to categorise the proteins into the two SRP groups.

```{r}
library(mixtools)

# find compound hydrophobicity score for each protein with a signal peptide
score_df <- add_compound_hydropathy_score(combined_df, AA_stringset, useSignalP = FALSE)
no_NA <- score_df %>%
    drop_na(compound_hydropathy)

mixed_model <- normalmixEM(no_NA$compound_hydropathy, k = 2)

# find intersection of two normal distributions
intersection <- uniroot(function(x) {
    dnorm(x, mean = mixed_model$mu[1], sd = mixed_model$sigma[1]) - dnorm(x, mean = mixed_model$mu[2], sd = mixed_model$sigma[2])
}, c(min(mixed_model$mu), max(mixed_model$mu)))$root
average <- mean(mixed_model$mu)

# plot distribution of scores with mixture model using ggplot2
# with an overlay of the two normal distributions
ggplot(no_NA, aes(x = compound_hydropathy)) +
    geom_histogram(aes(y = after_stat(density)), bins = 50, fill = "white", colour = "black") + 
    stat_function(fun = dnorm, args = list(mean = mixed_model$mu[1], sd = mixed_model$sigma[1]), colour = "red") +
    stat_function(fun = dnorm, args = list(mean = mixed_model$mu[2], sd = mixed_model$sigma[2]), colour = "green") +
    geom_vline(xintercept = intersection, colour = "blue", linetype = "dashed") +
    geom_vline(xintercept = average, colour = "purple", linetype = "dashed") +
    labs(x = "Compound Hydrophobicity Score", y = "Density", title = "Distribution of Compound Hydrophobicity Scores (Cerevisiae)") +
    theme_bw()
```

This is a plot of the distribution of compound hydrophobicity scores for proteins with signal peptides in N. Crassa.
The red and green lines are the two normal distributions that make up the mixture model, and the dashed lines are possible classification boundaries,
one at the intersection of the two normal distributions and one at the average of the two means.

Now we can use the binding motif or CRAC data to find out if Ssd-1 significantly targets one of these groups.
```{r}
# use mixed model and intersection to classify proteins into SRP groups, and make motif count groups
grouped_df <- score_df %>%
    drop_na(compound_hydropathy) %>%
    mutate(SRP_group = case_when(
        compound_hydropathy < intersection ~ "non_SRP",
        compound_hydropathy >= intersection ~ "SRP"
    )) %>%
    mutate(count_group = case_when(
        count_up_200 >= 1 ~ "1X 200nt",
        TRUE ~ "0X"
    )) %>%
    select(seqid, SRP_group, count_group, compound_hydropathy)

percentage_SRP_groups <- grouped_df %>%
    group_by(SRP_group) %>%
    summarise(percentage = n()/nrow(grouped_df) * 100)

# plot histogram of compound hydrophobicity scores for each SRP group
ggplot(grouped_df, aes(x = compound_hydropathy, fill = SRP_group)) +
    geom_histogram(aes(y = after_stat(density)), bins = 50, colour = "black") +
    facet_wrap(~count_group) +
    labs(x = "Compound Hydrophobicity Score", y = "Density", title = "Distribution of Compound Hydrophobicity Scores (Cerevisiae)") +
    theme_bw()

# plot bar chart of percentage of proteins in each SRP group
ggplot(grouped_df, aes(x = SRP_group, fill = SRP_group)) +
    geom_bar() +
    facet_wrap(~count_group) +
    labs(x = "SRP Group", y = "Count", title = "Number of Proteins in Each SRP Group (Cerevisiae)") +
    theme_bw()

# make contingency table
contingency_table <- grouped_df %>%
    count(count_group, SRP_group) %>%
    spread(SRP_group, n, fill = 0)

# run chi-squared test
chisq.test(contingency_table[,2:3])
```

The plot shows the distribution of hydrophobicity scores for the no CNYTCNYT motif and 1X 200nt CNYTCNYT motif groups.
The chi-squared test shows that the SRP groups are significantly dependent on the motif groups, indicating that the Ssd-1 binding motif is significantly enriched in the non-SRP group.